File: Go/zmisc/brackets.go
package main

import (
	"fmt"
	"strings"
)

func check(s string) bool {
	const (
		openBrackets  = "({["
		closeBrackets = ")}]"
	)
	stack := []byte{}
	for _, c := range s {
		openIdx := strings.IndexByte(openBrackets, byte(c))
		/*
			fmt.Println("openIdx:", openIdx)
			fmt.Println("c:", c)
			fmt.Println("byte(c):", byte(c))
		*/
		//closeIdx := strings.IndexByte(s, byte(c))
		if openIdx > -1 {
			stack = append(stack, byte(c))
		} else {
			slen := len(stack)
			if slen == 0 {
				return false
			}
			cc := stack[slen-1]
			stack = stack[:slen]
			//fmt.Println(stack)
			idx := strings.IndexByte(openBrackets, cc)
			if closeBrackets[idx] != byte(c) {
				return false
			}
		}
	}
	return true
}

func main() {
	s := "(){}[]"
	fmt.Println(check(s))
	s = "([)]"
	fmt.Println(check(s))
}

File: Java/zmisc/Brackets.java
import java.util.*;

public class Brackets {
    static ArrayList<Character> stack = new ArrayList<>();
    static final String openBrackets = "({[";
    static final String closeBrackets = ")}]";
    public static void main(String args[]) {
        var s = "(){}[]";
        for (var c : s.toCharArray()) {
            var ch = new Character(c);
            var openidx = openBrackets.indexOf(c);
            var closeidx = closeBrackets.indexOf(c);
            if (openidx > -1) {
                stack.add(ch);
            } else if (closeidx > -1) {
                if (stack.size() <= 0) {
                    System.out.println("stack 0");
                    return;
                }
                var lastIdx = stack.size() - 1;
                var cc = stack.get(lastIdx);
                stack.remove(lastIdx);
                var idx = openBrackets.indexOf(cc.charValue());
                if (closeBrackets.charAt(idx) != c) {
                    System.out.println("not valid!");
                    return;
                }
            }
        }
        System.out.println("valid!");
    }
}
